File: Cudone/Compiler/cudo.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Diagnostics;

namespace Compiler
{
    class Program
    {
        static void spt(int length)
        {
            string cout = "";
            int i = length;
            while(i != 0)
            {
                cout += "-";
                i --;
            }
            Console.WriteLine(cout);
        }

        static void hello(){
            Console.Write("Cudone - Programing Language Compiler\n");
            Console.Write("Copyright © Catrol, 2021\n");
            Console.Write("\n");
            Console.Write("View more info from the author with the address:\n");
            Console.Write("https://blog.catrol.cn\n");
        }

        static Queue<string> errs = new Queue<string>();
        static Queue<task> tasks = new Queue<task>();

        struct task{
            public string inputFN, outputFN;
            public bool binded;
            // FN - FileName
        }

        static string[] cmds = new string[4]
        {
            "-hello", "-i", "-o", "-d"
        };
        static string[] tips = new string[4]
        {
            "Error : code(0) : Please give at least 1 argument !!!",
            "Error : code(1) : 404 Command not found : ",
            "Error : code(2) : No input file after argument \"-i\" !!!",
            "Error : code(3) : No output file after argument \"-o\" !!!"
        };

        static void Processor(string[] args)
        {
            for(int i = 0; i < args.Length; ++ i)
            {
                int index = -1;
                for(int j = 0; j < cmds.Length; ++ j)
                {
                    if(cmds[j].Equals(args[i]))
                    {
                        index = j;
                        break;
                    }
                }
                switch(index)
                {
                    case 0:
                        Console.WriteLine("Hello Cudone !!!");
                        break;
                    case 1:
                        if(i + 1 < args.Length)
                        {
                            ++ i;
                            tasks.Enqueue(new task{
                                inputFN = args[i], outputFN = "output.cdbin", binded = false
                            });
                        }
                        else
                        {
                            errs.Enqueue(tips[2]);
                        }
                        break;
                    case 2:
                        if(i + 1 < args.Length)
                        {
                            ++ i;
                            int size = tasks.Count;
                            for(int j = 0; j < size; ++ j)
                            {
                                task item = (task)tasks.Dequeue();
                                if(!item.binded)
                                {
                                    item.outputFN = args[i];
                                    item.binded = true;
                                    tasks.Enqueue(item);
                                    break;
                                }
                                tasks.Enqueue(item);
                            }
                        }
                        else
                        {
                            errs.Enqueue(tips[3]);
                        }
                        break;
                    default:
                        errs.Enqueue($"{tips[1]}{args[i]}");
                        break;
                }
            }
        }

        static void Main(string[] args)
        {
            hello();
            if(args.Length != 0)
            {
                Processor(args);
            }
            else
            {
                Console.WriteLine(tips[0]);
            }
            if(errs.Count != 0)
            {
                Console.WriteLine();
                foreach(string item in errs)
                {
                    Console.WriteLine(item);
                }
                errs.Clear();
            }
            if(tasks.Count != 0)
            {
                Console.WriteLine();
                int index = 0;
                foreach(task item in tasks)
                {
                    Console.WriteLine($"task({index}) : -i {item.inputFN} - -o {item.outputFN}");
                    index ++;
                }
            }
            else
            {
                Console.WriteLine();
                Console.WriteLine("You added no tasks !!!");
                Console.WriteLine("Cudone Compiler exited.");
                return;
            }
            bool first = true;
            Console.WriteLine();
            while(true)
            {
                if(!first)
                {
                    Console.WriteLine("Please check in \"Y\" to confirm or \"n\" to cancel !");
                }
                else
                {
                    first = false;
                }
                Console.Write("Do you want to compile all of them now ? [Y/n] : ");
                string cos = Console.ReadLine();
                if(cos.Equals("Y"))
                {
                    break;
                }
                else if(cos.Equals("n"))
                {
                    Console.WriteLine();
                    Console.WriteLine("You have cancled all compile tasks !");
                    Console.WriteLine("Cudone Compiler exited.");
                    return;
                }
            }
            foreach(task item in tasks)
            {
                Translate(item.inputFN, item.outputFN);
            }
        }

        const int wordsnum = 43;

        static string[] keywords = new string[wordsnum]
        {
            "#导入 ", "使用库 ", " 主函数(", " 输出(\"", " 新建 ", " 返回 ", " <Cudone>", " 标准库;", "{字符串}", "{数}",
            "字符串", "数", "字符", "队列", "列表", "循环(", "做", "当(", "如果", "大于等于",
            "取模", "除", "乘", "加", "减", "^", "取地址->", "解引用->", "结构体 ", "布尔",
            "设为真", "设为假", "为真", "为假", "真", "假", "换行;", "小于等于", "等于", "并且",
            "大于", "小于", "或者"
        };

        static string[] tarwords = new string[wordsnum]
        {
            "#include ", "using namespace ", " main(", " printf(\"", " new ", " return ", " <bits/stdc++.h>", " std;", "%s", "%d",
            "string", "int", "char", "queue", "list", "for(", "do", "while(", "if", ">=",
            "%", "/", "*", "+", "-", "*", "&", "*", "struct ", "bool",
            " = true", " = false", " == true", " == false", "true", "false", "cout << endl;", "<=", "==", "&&",
            ">", "<", "||"
        };

        static void Translate(string ifn, string ofn)
        {
            if(File.Exists(ifn))
            {
                string cc = ReadAll(ifn);
                for(int i = 0; i < wordsnum; ++ i)
                {
                    cc = cc.Replace(keywords[i], tarwords[i]);
                }
                WriteIn($"{ofn}.cpp", cc);
                WriteIn($"{ofn}.com.sh", $"g++ {ofn}.cpp -o {ofn}.exe");
                Process process = new Process();
                process.StartInfo.FileName = "sh";
                process.StartInfo.Arguments = $"{ofn}.com.sh";
                process.Start();
                process.WaitForExit();
                process.Close();
                File.Delete($"{ofn}.cpp");
                File.Delete($"{ofn}.com.sh");
            }
        }
        
        static string ReadAll(string path)
        {
            string content;
            FileStream fs = new FileStream(path, FileMode.Open);
            StreamReader sr = new StreamReader(fs);
            content = sr.ReadToEnd();
            sr.Close(); sr.Dispose();
            fs.Close(); fs.Dispose();
            return content;
        }

        static void WriteIn(string path, string content)
        {
            if(File.Exists(path))
            {
                File.Delete(path);
            }
            FileStream fs = new FileStream(path, FileMode.OpenOrCreate);
            StreamWriter sr = new StreamWriter(fs);
            sr.Write(content);
            sr.Close(); sr.Dispose();
            fs.Close(); fs.Dispose();
        }
    }
}

File: Cudone/Compiler/cudo.cpp
#include <bits/stdc++.h>
#include <sys/stat.h>
#include <unistd.h>
#include <sstream>
#include <fstream>
#include <locale>
#define cmdNUM 4
#define rnrn spt(100);
using namespace std;
void spt(int length){
    while(length --) { printf("-%s", length == 0 ? "\n" : ""); }
}

bool IsPathExist(string path) {
    if (access(path.c_str(), 0) == F_OK) {
        return true;
    }
    return false;
}

bool IsFile(string path) {
    if (!IsPathExist(path)) {
        return false;
    }
    struct stat buffer;
    return (stat(path.c_str(), &buffer) == 0 && S_ISREG(buffer.st_mode));
}

string tips(int i){
    switch(i){
        case 0:
            return "Hello Cudone !!!";
        case 1:
            return "Error : code(1) : No input file after argument \"-i\" !!!";
        case 2:
            return "Error : code(2) : Command not found : ";
        case 3:
            return "Error : code(3) : No output file name after argument \"-o\" !!!";
        case 4:
            return "Error : code(4) : 404 not found File : ";
        default:
            return "\n";
    }
}

void hello(){
    printf("Cudone - Programing Language Compiler\n");
    printf("Copyright © Catrol, 2021\n");
    printf("\n");
    printf("View more info from the author with the address:\n");
    printf("https://blog.catrol.cn\n");
}

struct com_task{
    string inputFileName, outputFileName = "ouput.exe";
    bool binded = false;
};

const string cmds[cmdNUM] = {"-h", "-hello", "-i", "-o"};
queue<string> err;
queue<com_task> tasks;

void input(char* fileName){
    string fn = fileName;
    com_task ct;
    ct.inputFileName = fn;
    tasks.push(ct);
}

void output(char* tarName){
    string tn = tarName;
    int size = tasks.size();
    for(int i = 0; i < size; ++ i){
        com_task ct = tasks.front();
        if(!ct.binded){
            ct.outputFileName = tn;
            ct.binded = true;
            tasks.push(ct);
            tasks.pop();
            break;
        }else{
            tasks.push(ct);
            tasks.pop();
        }
    }
}

void processor(char* argv[], int num){
    for(int i = 1; i < num; ++ i){
        string argu = argv[i];
        int index = -1;
        for(int i = 0; i < cmdNUM; ++ i){
            if(argu == cmds[i]){
                index = i;
                break;
            }
        }
        switch(index){
            case 0:
                hello();
                break;
            case 1:
                printf("%s\n", tips(0).c_str());
                break;
            case 2:
                if(i + 1 == num){
                    err.push(tips(1));
                }else{
                    ++ i;
                    input(argv[i]);
                }
                break;
            case 3:
                if(i + 1 == num){
                    err.push(tips(3));
                }else{
                    ++ i;
                    output(argv[i]);
                }
                break;
            default:
                string info = tips(2);
                info += argu;
                err.push(info);
                break;
        }
    }
}

string& replace_all(string& src, const string& old_value, const string& new_value) {
    for (string::size_type pos(0); pos != string::npos; pos += new_value.length()) {
        if ((pos = src.find(old_value, pos)) != string::npos) {
            src.replace(pos, old_value.length(), new_value);
        }
        else break;
    }
    return src;
}

void writeTXT(string name, string content, bool cover){
    ofstream write(name, cover ? ios::trunc : ios::app);
    if (write.is_open()) {//如果成功的话        
        write << content << endl;  
        write.close();
    }
}

inline string read(const char* filename){
    ifstream file(filename);
    string content = "";
    if(file){
        stringstream buffer;
        buffer << file.rdbuf();
        content = buffer.str();
    }
    file.close();
    return content;
}

void compile(){
    int size = tasks.size();
    for(int i = 0; i < size; ++ i){
        com_task ct = tasks.front();
        if(IsFile(ct.inputFileName)){
            // string str = read(ct.inputFileName.c_str());
            // wifstream wifs(ct.inputFileName.c_str());
            // wstring s;
            // while(wifs.get(tmp))
            // {
            //     s.push_back(tmp);
            // }
            // wcout << fileContent << endl;
            string str;
            fstream reader(ct.inputFileName, fstream::in);
            fstream writter(ct.outputFileName, fstream::out);
            while(getline(reader, str)){
                replace_all(str, "#导入 ", "#include ");
                replace_all(str, " 主函数(", " main(");
                replace_all(str, "数", "int");
                replace_all(str, " 输出(\"", " printf(\"");
                replace_all(str, " 返回 ", " return ");
                for(int i = 0; i < str.length(); ++ i){
                    switch(str[i]){
                        case '<':
                            if(str[i + 1] == 'C' && str[i + 2] == 'u' && str[i + 3] == 'd' && str[i + 4] == 'o' && str[i + 5] == 'n' && str[i + 6] == 'e' && str[i + 7] == '>'){
                                str[i + 1] = '\0', str[i + 2] = '\0', str[i + 3] = '\0', str[i + 4] = '\0', str[i + 5] = '\0', str[i + 6] = '\0';
                                str.insert(i + 1, "bits/stdc++.h");
                            }
                            break;
                    }
                }
                cout << str << endl;
                writter << str << endl;
            }
            reader.close();
            writter.close();
            // wstring ss = to_wide_string(str);
            // wofstream fs;
            // fs.open(ct.outputFileName.c_str());
            // // locale utf8_locale(locale(), new utf8cvt<false>);
            // // fs.imbue(utf8_locale);
            // fs << ss.c_str() << endl;
            // fs.close();
            // ofstream out;
            // out.open(ct.outputFileName.c_str(), ios::out);
            // unsigned char smarker[3];
            // smarker[0] = 0xEF;
            // smarker[1] = 0xBB;
            // smarker[2] = 0xBF;
            // out << smarker << str;
            // out.close();
            // freopen(ct.outputFileName.c_str(), "w", stdout);
            // cout << to_byte_string(ss).c_str() << endl;
            // freopen("/dev/tty", "w", stdout);
            // freopen("/dev/tty", "r", stdin);
        }
        tasks.push(ct);
        tasks.pop();
        rnrn
    }
}

int main(int argc, char* argv[]){
    // locale china("chs");
    // wcin.imbue(china);//use locale object
    // wcout.imbue(china);
    if(argc == 1){
        err.push(tips(1));
    }else{
        processor(argv, argc);
    }
    if(!err.empty()){
        int size = err.size();
        for(int i = 0; i < size; ++ i){
            printf("%s", err.front().c_str());
            err.pop();
            printf("\n");
        }
    }
    if(!tasks.empty()){
        printf("Your compile tasks list produced !!!\n");
        rnrn
        int size = tasks.size();
        for(int i = 0; i < size; ++ i){
            com_task ct = tasks.front();
            printf("task(%d) :\n\t-i\t%s\n\t-o\t%s\n", i, ct.inputFileName.c_str(), ct.outputFileName.c_str());
            tasks.push(ct);
            tasks.pop();
        }
        rnrn
        bool loopTime = false;
        while(true){
            if(loopTime){
                printf("\nPlease check in \"Y\" to confirm or \"n\" to cancel !\n");
            }
            printf("Do you want to compile all of them now ? [Y/n] :");
            string in; getline(cin, in);
            if(in[0] == 'n' && in.length() == 1){
                rnrn
                printf("You have cancled all compile tasks !\n");
                printf("Cudone Compiler exited.\n");
                return 0;
            }else if(in[0] == 'Y' && in.length() == 1){
                break;
            }else{
                loopTime = true;
            }
        }
        rnrn
        compile();
    }
    return 0;
}
